Add rails 7.0 support and drop rails 6.1

kbrock opened the pull request:
Fixed includes and select issues

we're good to go for rails 7.0 with this

jrafanie commented on lib/active_record/virtual_attributes/virtual_fields.rb:
```diff
@@ -136,102 +143,70 @@ class Base
-              }
-              loaders
+        # preloader is called with virtual attributes - need to resolve
+        def call
```
where is `call` being called?

kbrock replied:
Great question.

Sure, we do explicitly call it in our specs: https://github.com/kbrock/activerecord-virtual_attributes/blob/allow_rails_7/spec/virtual_includes_spec.rb#L542

But most of the time, it is called directly by rails.

From what I can tell, this is mostly called directly from preload_associations via exec_query.

https://github.com/rails/rails/blob/7-0-stable/activerecord/lib/active_record/relation.rb#L825-L832

I did try patching in that caller, but when we ran the preloader directly, this had issues - as you know since you had to skip those 3 tests before (we were since able to unskip).

So this fixes 18 specs (probably 50+ uses) in virtual_includes_specs like:

```
expect(Author.includes(:nick_or_name)).to preload_values(:nick_or_name, author_name)
```

This `includes` value ends up in a Preloader. As you know, it is not valid. So we need to translate from the old includes values to the proper ones.

Now I did try and just update `@associations` value directly, but that only worked if there were no additional levels, so many tests were still failing.

This fix called allows us to get more complicated with our associations.
Coincidently, this recursive-ish solution is similar to the v6 changes where we call into self if the value changed. That was purely coincidental but brought me comfort.

jrafanie commented on lib/active_record/virtual_attributes/virtual_fields.rb:
```diff
@@ -136,102 +143,70 @@ class Base
+          def preloaders_for_reflection(reflection, reflection_records)
+            reflection_records.group_by do |record|
+              # begin virtual_attributes changes
+              needed_association = record.class.replace_virtual_fields(association)
```
is this for fixing a test fix?  This looks like it could "fix" live code with rails 7.  I'm not sure what we weren't testing previously with rails7 and master on virtual attributes to not need this. Is it just a performance fix only, so it doesn't break without it?

kbrock replied:
hmm... just tested.
We definitely had this before the preloader changes I did before.
So I think it did more in the past.

Now, it looks like this only fixes the 2 polymorphic tests.

```ruby
it "preloads through polymorphic (polymorphic => virtual_attribute)" do
  books = Book.includes(:author_or_bookmark => :total_books).load
  expect { expect(books.map { |b| b.author_or_bookmark.total_books }).to eq([3, 3, 3]) }.to_not make_database_queries
end
```

```
Association named 'total_books' was not found on Author; perhaps you misspelled it?
```

jrafanie replied:
ok, that makes sense.  `includes` not working wouldn't break cross repo tests but would break tests here that are specifically looking for query counts and the expected optimizations.  All good here.

jrafanie commented:
@kbrock in https://github.com/ManageIQ/activerecord-virtual_attributes/compare/master...jrafanie:activerecord-virtual_attributes:allow_rails_7_backup?expand=1 (the backup for the old PR), I was also able to delete `without_virtual_includes`, `apply_join_dependency`, and `calculate` from `virtual_fields.rb`.  Can you see if we can remove them?  See: https://github.com/ManageIQ/activerecord-virtual_attributes/commit/c4a34911f123bbb1cdcc204e9551cfdfc3c397dc

kbrock commented:
@jrafanie good catch. We can definitely deleted them. I thought I had.
they must have snuck back in during rebasing.

kbrock commented:
update:
- rebased
- added comments
- extracted a few PRs that are not directly responsible

update:
- removed extra changes to grouped_records - it had extra changes that were unnecessary. Think the changes to includes fixed this stuff as well.

jrafanie commented:
https://github.com/ManageIQ/manageiq/pull/22873 is updated to use this branch for rails 7.0 support, and cross repo is being run against it: https://github.com/ManageIQ/manageiq-cross_repo-tests/pull/871.  It's 💚 other than consumption but I'll take a look at that.  It's not a new problem.  It has been failing but because it always fails, I've looked at the other failures first.

```
    1) ManageIQ::Showback::DataRollup#engine update data_rollups should return value of the group with field
       Failure/Error: expect(data_rollup.get_group('CPU', 'average')).to eq([data_rollup.resource.metrics.for_time_range(data_rollup.start_time, data_rollup.end_time).average(:cpu_usage_rate_average).to_s, "percent"])
  
         expected: ["50.5", "percent"]
              got: [50.5, "percent"]
  
         (compared using ==)
       # ./spec/models/data_rollup_spec.rb:226:in `block (4 levels) in <top (required)>'
```

jrafanie commented:
@kbrock I think we're good to go on moving this forward. I think the test failure in consumption is unrelated to virtual attributes. Looking at this PR one more time.

kbrock commented:
update:
- rebase master
- updated comments in Association::Preloader - grouping by association rather than class (to avoid `return`ing from within the loop

kbrock commented:
This has changelog in it.
Do we think we will get in #150, or hold off? (I want to update the `CHANGELOG` to reflect)

jrafanie reviewed (approved):
We did a cross repo on this branch and it all passed 💚 

https://github.com/ManageIQ/manageiq-cross_repo-tests/pull/871
